server/src/main/java/com/cloud/network/NetworkModelImpl.java

```diff
@@ -593,13 +590,24 @@ public List<? extends Nic> getNics(long vmId) {
     @Override
     public String getNextAvailableMacAddressInNetwork(long networkId) throws InsufficientAddressCapacityException {
         NetworkVO network = _networksDao.findById(networkId);
-        String mac = _networksDao.getNextAvailableMacAddress(networkId, MACIdentifier.value());
-        if (mac == null) {
-            throw new InsufficientAddressCapacityException("Unable to create another mac address", Network.class, networkId);
+        Integer zoneIdentifier = MACIdentifier.value();
+        if (zoneIdentifier.intValue() == 0) {
+            zoneIdentifier = Long.valueOf(network.getDataCenterId()).intValue();
         }
+        String mac;
+        do {
+            mac = _networksDao.getNextAvailableMacAddress(networkId, zoneIdentifier);
+            if (mac == null) {
+                throw new InsufficientAddressCapacityException("Unable to create another mac address", Network.class, networkId);
+            }
+        } while(isMACUnique(mac));
```

Comment:
@DaanHoogland I may be wrong but shouldn't this condition be reversed?

Reply:
that, or the name is wrong. I'll have a look, tnx

Reply:
```suggestion
        } while(! isMACUnique(mac));
```